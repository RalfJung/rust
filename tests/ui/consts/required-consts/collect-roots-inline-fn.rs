//@revisions: noopt opt
//@ build-fail
//@[noopt] compile-flags: -Copt-level=0
//@[opt] compile-flags: -O

//! In optimized builds, the functions in this crate are all marked "inline" so none of them become
//! collector roots. Ensure that we still evaluate the constants.
#![crate_type = "lib"]

struct Zst<T>(T);
impl<T> Zst<T> {
    const ASSERT: () = if std::mem::size_of::<T>() != 0 {
        panic!(); //~ERROR: evaluation of `Zst::<u32>::ASSERT` failed
    };
}

fn f<T>() {
    Zst::<T>::ASSERT;
}

pub fn g() {
    f::<u32>()
}
