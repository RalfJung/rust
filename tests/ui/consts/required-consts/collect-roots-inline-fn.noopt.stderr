error[E0080]: evaluation of `Zst::<u32>::ASSERT` failed
  --> $DIR/collect-roots-inline-fn.rs:13:9
   |
LL |         panic!();
   |         ^^^^^^^^ the evaluated program panicked at 'explicit panic', $DIR/collect-roots-inline-fn.rs:13:9
   |
   = note: this error originates in the macro `$crate::panic::panic_2015` which comes from the expansion of the macro `panic` (in Nightly builds, run with -Z macro-backtrace for more info)

note: erroneous constant encountered
  --> $DIR/collect-roots-inline-fn.rs:18:5
   |
LL |     Zst::<T>::ASSERT;
   |     ^^^^^^^^^^^^^^^^

note: the above error was encountered while instantiating `fn f::<u32>`
  --> $DIR/collect-roots-inline-fn.rs:22:5
   |
LL |     f::<u32>()
   |     ^^^^^^^^^^

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0080`.
